File: storage/src/main/java/pl/expensive/storage/Transaction.java
package pl.expensive.storage;

import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;

import org.threeten.bp.Instant;
import org.threeten.bp.LocalDateTime;
import org.threeten.bp.ZoneId;
import org.threeten.bp.ZoneOffset;

import java.math.BigDecimal;
import java.util.UUID;

@AutoValue
public abstract class Transaction {
    public abstract UUID uuid();
    public abstract UUID wallet();
    public abstract BigDecimal amount();
    public abstract Currency currency();
    // time in millis
    public abstract long date();
    @Nullable
    public abstract String description();

    public static Transaction create(UUID uuid, UUID wallet, BigDecimal amount, Currency currency, LocalDateTime date, String desc) {
        return new AutoValue_Transaction(uuid, wallet, amount, currency, toMillisUTC(date), desc);
    }

    public static Transaction create(UUID uuid, UUID wallet, BigDecimal amount, Currency currency, long time, String desc) {
        return new AutoValue_Transaction(uuid, wallet, amount, currency, time, desc);
    }

    public static Transaction withdrawal(UUID wallet, BigDecimal amount, Currency currency, String desc) {
        return create(UUID.randomUUID(), wallet, amount.negate(), currency, toMillisUTC(LocalDateTime.now()), desc);
    }

    public static Transaction deposit(UUID wallet, BigDecimal amount, Currency currency, String desc) {
        return create(UUID.randomUUID(), wallet, amount, currency, toMillisUTC(LocalDateTime.now()), desc);
    }

    public LocalDateTime toLocalDate() {
        return Instant.ofEpochMilli(date())
                .atZone(ZoneId.of("UTC"))
                .toLocalDateTime();
    }

    private static long toMillisUTC(LocalDateTime date) {
        return date.toInstant(ZoneOffset.UTC).toEpochMilli();
    }
}

File: storage/src/main/java/pl/expensive/storage/Transaction.kt
package pl.expensive.storage

import android.os.Parcel
import android.os.Parcelable
import java.math.BigDecimal
import java.util.*

fun withdrawal(uuid: UUID = UUID.randomUUID(),
               amount: BigDecimal,
               currency: Currency,
               time: Long = Date().time,
               desc: String = "",
               category: Category? = null) =
        Transaction(uuid, amount.negate(), currency, time, desc, category)

data class Transaction(val uuid: UUID,
                       val amount: BigDecimal,
                       val currency: Currency,
                       val date: Long, // time in millis
                       val description: String,
                       var category: Category? = null) : Parcelable {

    constructor(parcelIn: Parcel) : this(
            parcelIn.readString().toUUID(),
            parcelIn.readString().asBigDecimal(),
            parcelIn.readParcelable(Currency.CREATOR)!!,
            parcelIn.readLong(),
            parcelIn.readString(),
            parcelIn.readParcelable(Category.CREATOR))

    override fun writeToParcel(dest: Parcel, flags: Int) {
        with(dest) {
            writeString(uuid.toString())
            writeString(amount.toString())
            writeParcelable(currency, 0)
            writeLong(date)
            writeString(description)
            writeParcelable(category, 0)
        }
    }

    override fun describeContents() = 0

    // TODO: Maybe make separate functions instead of companion object for factory methods?
    companion object {
        @JvmField @Suppress("unused")
        val CREATOR = createParcel(::Transaction)
    }
}
